### DISCUSSION TITLE
Wazuh Vulnerability Detector

### DISCUSSION THREAD

**Discussion by kiaraho** [Security-Onion-Solutions/securityonion#8218] [Unsupported Versions] [answered]
Has anyone managed to get the Wazuh `vulnerability-detector` module working in Security Onion? The suggested answer in #4875 does not seem to work in a distributed environment.

I made the [configuration changes here](https://documentation.wazuh.com/3.13/user-manual/capabilities/vulnerability-detection/running_vu_scan.html) on a sensor monitoring endpoints with outdated software packages, but did not receive any expected alerts.

**Comment by dougburks** [comment]
Have you checked the Wazuh server logs on the sensor or the Wazuh agent logs on the endpoints for additional clues?

**Reply by kiaraho** [reply]
@dougburks, I checked `/nsm/wazuh/queue/vulnerabilities/cve.db` on the sensor, but the `AGENTS` and `NVD_CVE_MATCH` tables are empty. The `VULNERABILITIES` table has 53,289 entries so the NVD database was successfully populated.

Pivoting to the `syscollector` host dataset in Kibana, I see the sensor is collecting system inventory info from the agents as expected. This info includes the `program.name`, `program.vendor` and `program.version` fields, that I'm assuming are used to do the NVD checks.

Not sure where else to look.

**Comment by jctello** [comment, marked as answer]
Hi @kiaraho, 
The `VULNERABILITIES` table does not only reflect NVD data, instead `NVD_CVE` will do that, and you may find this table is still empty:
```
sqlite3 /nsm/wazuh/queue/vulnerabilities/cve.db  'select * from NVD_CVE limit 1;'
```
in the Wazuh logs at `/nsm/wazuh/logs/ossec.log` you may have noticed a message similar to:
```
wazuh-modulesd:vulnerability-detector: WARNING: (5522): There was no valid response to 'https://nvd.nist.gov/feeds/json/cve/1.0/nvdcve-1.0-2010.meta' after '3' attempts.
```
Security Onion uses Wazuh 3.13.1, during the lifetime of this version the NVD 1.0 CVE feed was deprecated which rendered the URL used by this version of Wazuh outdated. This was reported in https://github.com/wazuh/wazuh/issues/6053 and fixed for Wazuh 3.13.2.

In order to use vulnerability detection on this version you must rely on an offline update of the NVD DB as explained here: https://documentation.wazuh.com/3.13/user-manual/capabilities/vulnerability-detection/offline_update.html#national-vulnerability-database however, [as mentioned in the conversation](https://github.com/wazuh/wazuh/issues/6053#issuecomment-694989665) of the issue above, in version 3.13.1 Wazuh was not able to work with compressed files for custom URLs so the way to get Vulnerability Detection to work on the current version of Security Onion is:
 - Download the offline update script for NVD
 - Use the script to download the feed files
 - Uncompress the feed files
 - Configure Wazuh to use these files

This can be achieved with the following commands:
```
mkdir /nsm/wazuh/etc/nvd-db
curl -O https://raw.githubusercontent.com/wazuh/wazuh/master/tools/vulnerability-detector/nvd-generator.sh
chmod +x nvd-generator
./nvd-generator.sh 2002 /nsm/wazuh/etc/nvd-db/
gunzip /nsm/wazuh/etc/nvd-db/nvd-feed*
```
Ensure your Wazuh configuration at `/nsm/wazuh/etc/ossec.conf` includes a vulnerability configuration like this one:
```
 <vulnerability-detector>
    <enabled>yes</enabled>
    <interval>5m</interval>
    <ignore_time>6h</ignore_time>
    <run_on_start>yes</run_on_start>


    <provider name="redhat">
        <enabled>yes</enabled>
        <update_from_year>2010</update_from_year>
        <update_interval>1h</update_interval>
    </provider>

    <provider name="nvd">
        <enabled>yes</enabled>
        <update_from_year>2010</update_from_year>
        <path>/var/ossec/etc/nvd-db/.*json$</path>
        <update_interval>1h</update_interval>
    </provider>

</vulnerability-detector>
```
And of course restart Wazuh with `so-wazuh-restart` for the changes to take effect.

**Reply by kiaraho** [reply]
@jctello, this is the error I got in the `/nsm/wazuh/logs/ossec.log` file after following your recommendations:
```
2022/07/06 15:01:49 wazuh-modulesd:vulnerability-detector: INFO: (5400): Starting 'National Vulnerability Database' database update.
2022/07/06 15:01:49 wazuh-modulesd:vulnerability-detector: ERROR: (5551): Invalid multi_path '/nvd-db': 'No such file or directory'
2022/07/06 15:01:49 wazuh-modulesd:vulnerability-detector: ERROR: (5513): CVE database could not be updated.
2022/07/06 15:01:49 wazuh-modulesd:vulnerability-detector: INFO: (5431): Starting vulnerability scan.
2022/07/06 15:01:49 wazuh-modulesd:vulnerability-detector: ERROR: (5582): Unavailable vulnerabilities at the NVD database. The scan is aborted.
``` 
Here are the current contents of the `/nvd-db` directory:
```
-rw-r--r-- 1 root root  21M Jul  6 14:21 nvd-feed2002.json
-rw-r--r-- 1 root root 5.8M Jul  6 14:21 nvd-feed2003.json
-rw-r--r-- 1 root root  12M Jul  6 14:21 nvd-feed2004.json
-rw-r--r-- 1 root root  19M Jul  6 14:21 nvd-feed2005.json
-rw-r--r-- 1 root root  30M Jul  6 14:21 nvd-feed2006.json
-rw-r--r-- 1 root root  28M Jul  6 14:21 nvd-feed2007.json
-rw-r--r-- 1 root root  32M Jul  6 14:21 nvd-feed2008.json
-rw-r--r-- 1 root root  31M Jul  6 14:21 nvd-feed2009.json
-rw-r--r-- 1 root root  34M Jul  6 14:21 nvd-feed2010.json
-rw-r--r-- 1 root root  34M Jul  6 14:21 nvd-feed2011.json
-rw-r--r-- 1 root root  39M Jul  6 14:21 nvd-feed2012.json
-rw-r--r-- 1 root root  44M Jul  6 14:21 nvd-feed2013.json
-rw-r--r-- 1 root root  39M Jul  6 14:21 nvd-feed2014.json
-rw-r--r-- 1 root root  39M Jul  6 14:21 nvd-feed2015.json
-rw-r--r-- 1 root root  50M Jul  6 14:21 nvd-feed2016.json
-rw-r--r-- 1 root root  71M Jul  6 14:22 nvd-feed2017.json
-rw-r--r-- 1 root root  74M Jul  6 14:22 nvd-feed2018.json
-rw-r--r-- 1 root root  84M Jul  6 14:22 nvd-feed2019.json
-rw-r--r-- 1 root root 107M Jul  6 14:22 nvd-feed2020.json
-rw-r--r-- 1 root root 117M Jul  6 14:22 nvd-feed2021.json
-rw-r--r-- 1 root root  38M Jul  6 14:22 nvd-feed2022.json
```

**Reply by jctello** [reply]
Hello @kiaraho, 
I forgot to account for the fact that on Security Onion the Wazuh manager is run in a docker container so it cannot see the `/nvd-db` folder. So instead I recommend downloading the NVD DB to `/nsm/wazuh/etc/nvd-db/` and using the `/var/ossec/etc/nvd-db` path in the configuration. 

I will update my answer above to correct for this.

**Reply by kiaraho** [reply]
That got rid of the 5551 and 5513 errors. The 5582 error is still present after the `so-wazuh-restart`.
```
2022/07/06 16:02:33 wazuh-modulesd:vulnerability-detector: INFO: (5400): Starting 'National Vulnerability Database' database update.
2022/07/06 16:02:33 wazuh-modulesd:vulnerability-detector: INFO: (5430): The update of the 'National Vulnerability Database' feed finished successfully.
2022/07/06 16:02:33 wazuh-modulesd:vulnerability-detector: INFO: (5431): Starting vulnerability scan.
2022/07/06 16:02:33 wazuh-modulesd:vulnerability-detector: ERROR: (5582): Unavailable vulnerabilities at the NVD database. The scan is aborted.
```

**Reply by jctello** [reply]
I ran into this when the path was correct but the files were not yet uncompressed, I just noticed I didn't update that path in my initial answer.

This may also be a race condition where vulnerability detector keeps the status of the previous error and does not really update the database. So I recommend ensuring that the contents of `/nsm/wazuh/etc/nvd-db` are indeed uncompressed and running `so-wazuh-restart` once more.

**Reply by kiaraho** [reply]
I just moved the JSON files I had downloaded earlier from the `/nvd-db` directory to `/nsm/wazuh/etc/nvd-db`. I can confirm the JSON files are indeed uncompressed.
```
$ sudo file /nsm/wazuh/etc/nvd-db/nvd-feed2022.json
/nsm/wazuh/etc/nvd-db/nvd-feed2022.json: UTF-8 Unicode text, with very long lines, with CRLF line terminators
```
I ran `so-wazuh-restart` several times and even rebooted the sensor. Same 5582 error.

**Reply by jctello** [reply]
I'm get a slightly different file encoding:
```
sudo file /nsm/wazuh/etc/nvd-db/nvd-feed2022.json
/nsm/wazuh/etc/nvd-db/nvd-feed2022.json: ASCII text, with CRLF line terminators
```
Can you verify the contents of the file?:
```
head /nsm/wazuh/etc/nvd-db/nvd-feed2022.json
{
  "CVE_data_type" : "CVE",
  "CVE_data_format" : "MITRE",
  "CVE_data_version" : "4.0",
  "CVE_data_numberOfCVEs" : "8301",
  "CVE_data_timestamp" : "2022-07-05T07:00Z",
  "CVE_Items" : [ {
    "cve" : {
      "data_type" : "CVE",
      "data_format" : "MITRE",
```
It would also be worth a try to delete the db and have it be created from scratch:
```
rm /nsm/wazuh/queue/vulnerabilities/cve.db && so-wazuh-restart && tail -f /nsm/wazuh/logs/ossec.log  | grep vuln
```
For this I got over a span of 10 minutes:
```
2022/07/06 17:49:23 wazuh-modulesd:vulnerability-detector: INFO: (5400): Starting 'Red Hat Enterprise Linux' database update.
2022/07/06 17:50:06 wazuh-modulesd:vulnerability-detector: INFO: (5430): The update of the 'Red Hat Enterprise Linux' feed finished successfully.
2022/07/06 17:50:08 wazuh-modulesd:vulnerability-detector: INFO: (5400): Starting 'National Vulnerability Database' database update.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5430): The update of the 'National Vulnerability Database' feed finished successfully.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5431): Starting vulnerability scan.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5450): Analyzing agent '000' vulnerabilities.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: WARNING: (5575): Unavailable vulnerability data for the agent '000' OS. Skipping it.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5450): Analyzing agent '001' vulnerabilities.
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5471): Finished vulnerability assessment for agent '001'
2022/07/06 17:57:51 wazuh-modulesd:vulnerability-detector: INFO: (5472): Vulnerability scan finished.
```
it's normal that agent 000 (the manager) running on docker has no valid Vulnerability feed source. However the agent running on the host machine does get vulnerabilities:
![image](https://user-images.githubusercontent.com/36071202/177614947-9ab0f922-2208-47cd-b05b-7f86a7167a98.png)

**Reply by kiaraho** [reply]
I have a similar first 10 lines in my `nvd-feed2022.json` file.
```
sudo head /nsm/wazuh/etc/nvd-db/nvd-feed2022.json
{
  "CVE_data_type" : "CVE",
  "CVE_data_format" : "MITRE",
  "CVE_data_version" : "4.0",
  "CVE_data_numberOfCVEs" : "8329",
  "CVE_data_timestamp" : "2022-07-06T07:00Z",
  "CVE_Items" : [ {
    "cve" : {
      "data_type" : "CVE",
      "data_format" : "MITRE",
```
After some more digging, I found a typo in the `<path>` section of my `/nsm/wazuh/etc/ossec.conf` file. I had `../*json$` instead of `../.*json$`. Fixed it and ran `so-wazuh-restart`. Everything seems to be working now. Thanks!
```
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: INFO: (5431): Starting vulnerability scan.
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: INFO: (5450): Analyzing agent '000' vulnerabilities.
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: WARNING: (5575): Unavailable vulnerability data for the agent '000' OS. Skipping it.
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: INFO: (5450): Analyzing agent '001' vulnerabilities.
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: WARNING: (5575): Unavailable vulnerability data for the agent '001' OS. Skipping it.
2022/07/06 19:11:41 wazuh-modulesd:vulnerability-detector: INFO: (5450): Analyzing agent '002' vulnerabilities.
```

**Reply by jctello** [reply]
Great, thank you as well.
It was an interesting case to work. 
And in all honesty that was in my original answer, I'm used to using **glob** expressions for paths but this one used a **regex** so it was then looking for instances of `/` instead of any character.
I fixed it at some point and forgot to mention it.
